dali/kernels/reduce/reduce.h

```diff
@@ -0,0 +1,453 @@
+  template <typename Acc, typename Addend>
+  DALI_HOST_DEV DALI_FORCEINLINE
+  void operator()(Acc &acc, const Addend &val) const noexcept {
+    acc += val;
+  }
+};
+
+}  // namespace reductions
+
+
+constexpr int kTreeReduceThreshold = 32;
+
+template <int static_stride, typename Dst, typename Src, typename Preprocessor, typename Reduction>
+void reduce1D_stride(Dst &reduced, const Src *data, int64_t dynamic_stride, int64_t n,
+                     const Preprocessor &P, const Reduction &R) {
+  const int64_t stride = static_stride < 0 ? dynamic_stride : static_stride;
+  if (n > kTreeReduceThreshold) {
+    int64_t m = n >> 1;
+    Dst tmp1 = 0, tmp2 = 0;
+    // reduce first half and accumulate
+    reduce1D_stride<static_stride>(tmp1, data, stride, m, P, R);
+    // reduce second half and accumulate
+    reduce1D_stride<static_stride>(tmp2, data + m * stride, stride, n - m, P, R);
+    R(tmp1, tmp2);
+    R(reduced, tmp1);
+  } else {
+    // reduce to a temporary
+    Dst tmp = 0;
+    for (int64_t i = 0; i < n; i++)
+       R(tmp, P(data[i * stride]));
+    // accumulate in target value
+    R(reduced, tmp);
+  }
+}
+
+template <typename Dst, typename Src, typename Preprocessor, typename Reduction>
+void reduce1D(Dst &reduced, const Src *data, int64_t stride, int64_t n,
+              const Preprocessor &P, const Reduction &R) {
+  VALUE_SWITCH(stride, static_stride, (1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 16),
+    (reduce1D_stride<static_stride>(reduced, data, static_stride, n, P, R);),
+    (reduce1D_stride<-1>(reduced, data, stride, n, P, R);)
+  );  // NOLINT
+}
+
+namespace detail {
+template <typename Backend, typename T>
+struct StridedTensor {
+  T *data = nullptr;
+  int dim() const noexcept { return size.size(); }
+  SmallVector<int, 6> stride, size;
+};
+
+/// @brief Reduces a strided tensor slice to a scalar
+template <typename Dst, typename Src, typename Preprocessor, typename Reduction>
+void reduce(Dst &reduced, const StridedTensor<StorageCPU, Src> &in,
+            const Preprocessor &P, const Reduction &R,
+            int axis, int64_t extent, int64_t offset) {
+  int64_t stride = in.stride[axis];
+  if (axis == in.dim() - 1) {
+    Dst tmp = 0;
+    reduce1D(tmp, in.data + offset, stride, in.size[axis], P, R);
+    R(reduced, tmp);
+  } else {
+    int64_t sub_v = volume(in.size.begin() + axis + 1, in.size.end());
+    if (sub_v == 0)
+      sub_v = 1;
+    if (extent >= 2 && extent * sub_v > kTreeReduceThreshold) {
+      Dst tmp1 = 0, tmp2 = 0;
+      int64_t mid = extent / 2;
+      reduce(tmp1, in, P, R, axis, mid, offset);
+      reduce(tmp2, in, P, R, axis, extent - mid, offset + mid * stride);
+      R(tmp1, tmp2);
+      R(reduced, tmp1);
+    } else {
+      for (int64_t i = 0; i < extent; i++) {
+        Dst tmp = 0;
+        reduce(tmp, in, P, R, axis + 1, in.size[axis + 1], offset + i * stride);
+        R(reduced, tmp);
+      }
+    }
+  }
+}
+
+/// @brief Reduces a strided tensor to a scalar
```

Comment:
I thought we agreed on a documentation style... https://github.com/NVIDIA/DALI/blob/master/STYLE_GUIDE.md#documentation

Reply:
I haven't. I was on vacation when that happened.

Reply:
Still, majority has voted and you should submit.

Reply:
I'm more than willing to submit - a change to that guide.